lib/coinbase/address/wallet_address.rb

```diff
@@ -64,20 +64,26 @@ def transfer(amount, asset_id, destination)
     # @param to_asset_id [Symbol] The ID of the Asset to trade to. For Ether, :eth, :gwei, and :wei are supported.
     # @return [Coinbase::Trade] The Trade object.
     def trade(amount, from_asset_id, to_asset_id)
-      from_asset = Asset.fetch(network_id, from_asset_id)
-      to_asset = Asset.fetch(network_id, to_asset_id)
-
-      validate_can_trade!(amount, from_asset)
-
-      trade = create_trade(amount, from_asset, to_asset)
+      ensure_can_sign!
+      ensure_sufficient_balance!(amount, from_asset_id)
```

Comment:
This is dramatically more readable in my opinion and are Address level concerns.

Whereas the different payloads belonging to a trade and fetching the assets needed for the trade, etc... are Trade level concerns.

Reply:
We do something very similar in staking. I wonder, should we be moving this logic into the Server and returning proper errors instead? Im guessing that is more inline with your aspirations, but curious your thoughts.

Reply:
Big +1 there, trying to decouple and break these down piece by piece to make the SDK layer light-weight.